### PR TITLE
Anomaly Detection Model on Time Series data using Isolation Forest

### LINKED ISSUE #6773: [Machine Learning] Anomaly Detection Model on Time Series data using Isolation Forest
#### Proposed title of article
[Machine Learning] Anomaly Detection Model on Time Series data using Isolation Forest

#### Proposed article introduction
A time series is a series of data points indexed in time order. Most commonly, a time series is a sequence taken at successive equally spaced points in time. Time series forecasting is the use of a model to predict future values based on previously observed values. Timeseries dataset may have anomalies or outliers that may lead to inconsistent results. Anomalies/outliers are observations or data points that deviate from a dataset’s normal behavior.

Anomalies in data are also called standard deviations, outliers, noise, novelties, and exceptions. Anomaly detection is the process of finding outliers in a given dataset. Outliers are the data objects that stand out amongst other objects in the dataset and do not conform to the normal behavior in a dataset. Anomaly detection identifies rare events, items, or observations that are suspicious because they differ significantly from standard behaviors or patterns. 

Isolation forest is a machine learning algorithm for anomaly detection. It's an unsupervised learning algorithm that identifies anomalies by isolating outliers in the data. Isolation Forest is based on the Decision Tree algorithm.

#### Key takeaways 
1. Dataset preprocessing in time series dataset.
2. Time series analysis and decomposition.
3. Plotting trends, cyclic and seasonality graphs.
4. Implementing Anamoly detection using Isolation Forest.
5. Anomalies visualization using Plotly Express.

#### Article quality
This tutorial will explain time series analysis and decomposition in detail. We will discuss the decomposable components such as trends, cyclic, and seasonality components. The tutorial will also discuss the two types of anomalies: Contextual and global anomalies. We will deal with contextual anomalies in detail and implement an Isolation Forest algorithm to detect the anomalies. Finally, we will plot a visualization graph to show all the anomalies in the dataset. This will help to inconsistent results in model training and forecasting.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by collinskirui223**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6773

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Bayler** [comment]
@collinskirui223 Please make sure your avatar image is 600px by 400px and not 600px by 600px

**Comment by collinskirui223** [comment]
I have resized the avatar image @Bayler.
Thanks

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+We will also be able to gain valuable insights from the dataset. The sudden spikes and drops may lead to inconsistent results during forecasting. 
+
+- It identifies noise in the dataset.
+Noise is unwanted and erroneous data points. Noisy data has meaningless information that may corrupt the time series model in training. Removing the noise will ensure we have a high-quality dataset.
+
+- It helps in identifying failures/malfunctioning in applications and devices.
+The model identifies unexpected changes in time-series data that record devices and applications' performance. The unexpected changes may be associated with device failure or malfunctioning.
+
+- It helps in network intrusion and network anomaly detection.
+Models that monitor the network traffic will detect sudden changes in network traffic. These sudden changes may be due to cyber-attacks and other unauthorized access.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:72]:
```suggestion
- It helps with fraud detection. Anomaly detection will identify unusual and suspicious events in time series. Anomaly detection models are applied in banks and other financial institutions to detect fraud. The anomaly detection model will identify suspicious activities and transactions.

- It detects sudden spikes and drops in the time series dataset. Anomaly detection will explain the sudden spikes and drops in the dataset. We will also be able to gain valuable insights from the dataset. The sudden spikes and drops may lead to inconsistent results during forecasting. 

- It identifies noise in the dataset. Noise is unwanted and erroneous data points. Noisy data has meaningless information that may corrupt the time series model in training. Removing the noise will ensure we have a high-quality dataset.

- It helps in identifying failures/malfunctioning in applications and devices. The model identifies unexpected changes in time-series data that record devices and applications' performance. The unexpected changes may be associated with device failure or malfunctioning.

- It helps in network intrusion and network anomaly detection. Models that monitor the network traffic will detect sudden changes in network traffic. These sudden changes may be due to cyber-attacks and other unauthorized access.
```

Comment (Bayler) [reply content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:72]: As a rule of thumb, make sure your paragraphs are at least 3 lines long.

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+### Table of contents
+- [Prerequisites](#prerequisites)
+- [Why perform anomaly detection?](#why-perform-anomaly-detection)
+- [Dataset preparation](#dataset-preparation)
+- [Dataset resampling](#dataset-resampling)
+- [Adding new columns](#adding-new-columns)
+- [Plotting line charts](#plotting-line-charts)
+- [Plotting a detailed line chart](#plotting-a-detailed-line-chart)
+- [Building the anomaly detection using Isolation Forest](#building-the-anomaly-detection-using-isolation-forest)
+- [Initialiaze the Isolation Forest algorithm](#initialiaze-the-isolation-forest-algorithm)
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:38]:
```suggestion
- [Initialize the Isolation Forest algorithm](#initialize-the-isolation-forest-algorithm)
```

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+- It helps in identifying failures/malfunctioning in applications and devices.
+The model identifies unexpected changes in time-series data that record devices and applications' performance. The unexpected changes may be associated with device failure or malfunctioning.
+
+- It helps in network intrusion and network anomaly detection.
+Models that monitor the network traffic will detect sudden changes in network traffic. These sudden changes may be due to cyber-attacks and other unauthorized access.
+
+### Dataset preparation
+We will prepare the dataset that the anomaly detection model will use. We will use the New York Taxi dataset. 
+
+The dataset shows the number of taxi rides for six months. It is in 30 minutes intervals.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:77]:
```suggestion
We will prepare the dataset that the anomaly detection model will use. We will use the New York Taxi dataset. The dataset shows the number of taxi rides for six months. It is in 30 minutes intervals. 
```

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+df
+```
+Output:
+
+![Rows and columns](/engineering-education/anomaly-detection-model-on-time-series-data-using-isolation-forest/taxi-dataset.png)
+
+The dataset has `timestamp` and `value` columns.
+
+The `timestamp` column shows the time of recording.
+The `value` column contains the number of taxi rides.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:104]:
```suggestion
The dataset has `timestamp` and `value` columns. The `timestamp` column shows the time of recording. The `value` column contains the number of taxi rides.
```

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+Output:
+
+![New columns](/engineering-education/anomaly-detection-model-on-time-series-data-using-isolation-forest/new-columns.png)
+
+### Plotting line charts
+We will plot two line charts. The first line chart will show the number of taxi rides during each day of the week. 
+
+The second line chart will show the number of taxi rides during each hour of the day. 
+
+We will plot the line chart using Matplotlib. Let's import Matplotlib.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:167]:
```suggestion
We will plot two line charts. The first line chart will show the number of taxi rides during each day of the week. The second line chart will show the number of taxi rides during each hour of the day. We will plot the line chart using Matplotlib. 

Let's import Matplotlib.
```

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+fig.show()
+```
+Detailed line chart output:
+
+![Detailed line chart](/engineering-education/anomaly-detection-model-on-time-series-data-using-isolation-forest/detailed-line-chart.png)
+
+We have prepared our dataset and visualized the data points. Let's start building the anomaly detection model using an Isolation Forest.
+
+### Building the anomaly detection using Isolation Forest
+Isolation Forest is an unsupervised algorithm for anomaly detection. It does not require a labeled time-series dataset to identify anomalies.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:225]: Please add an external link to the Isolation Forest documentation.

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+Isolation Forests(IF), similar to [Random Forests](/engineering-education/introduction-to-random-forest-in-machine-learning/), are built based on decision trees. 
+
+An Isolation Forest algorithm randomly selects features in the time series dataset. It then sub-samples data points using a tree-based structure. 
+
+Isolation Forest builds multiple decision trees so that each leaf of the decision tree will isolate the data points in the time series dataset. Each leaf of the tree isolates exactly one observation from the data set. 
+
+The data point can either be an anomaly or not. Let's initialize the Isolation Forest algorithm.
+
+#### Initialize the Isolation Forest algorithm
+use this code:
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:241]: You can't just write, "use this code". Make sure your sentences appear complete.

File: content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+We will use Plotly Express to visualize all the anomalies in our dataset. Run this code:
+
+```python
+fig = px.scatter(df.reset_index(), x='timestamp', y='value', color='outliers', hover_data=['weekday'], title='TAXI DRIVES')
+fig.update_xaxes(
+    rangeslider_visible=True,
+)
+fig.show()
+```
+The above Ploty Express diagram shows the anomalous data points/outliers and the normal data points.
```
Comment (Bayler) [inline content/articles/anomaly-detection-model-on-time-series-data-using-isolation-forest/index.md:296]: Where's the diagram that you're referring to?

File: content/authors/collins-kirui/index.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
+title: Collins Kirui
+type: authors
+github: https://github.com/collinskirui223
+images:
+  - url: /engineering-education/authors/collins-kirui/avatar.jpg 
+---
```
Comment (Bayler) [inline content/authors/collins-kirui/index.md:7]: Please refer to this [documentation](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/UPLOAD_INSTRUCTIONS.md) and add your skillset into your bio index.md file.

**Comment by Bayler** [comment]
@collinskirui223 Please work on the revisions before we can move on for final review.

**Comment by collinskirui223** [comment]
Hello @Bayler, thanks for the review and the suggestions. I have implemented all your suggestions.

**Comment by hectorkambow** [comment]
@collinskirui223 

cc @Bayler  

This article seems rather similar in process to [this one](https://www.section.io/engineering-education/anomaly-detection-model-on-time-series-data/)

**Comment by jamesomina99** [comment]
@hectorkambow, In both the articles we are building a common time series model to explain the concept of anomaly detection, but we are using different libraries which brings uniqueness. 

The other author is using Facebook Prophet but I am using Isolation Forest. Kindly reconsider reopening this PR.

**Comment by collinskirui223** [comment]
@hectorkambow, this is my first PR, so when doing research I decided to build a simple time series model that is commonly used that is suitable for readers (especially beginners). What brings the uniqueness between the two articles is that I am using the Isolation Forest algorithm to build the model but the other author is using Facebook Prophet.

These are different libraries, even though the process of building the time series model is the same. I have also explained different concepts that are not in the other article.

Thanks. 
cc @Bayler

**Comment by hectorkambow** [comment]
I think I was reading the other one by mistake - they are both in the PR. 


Both index files are include - please double check

**Comment by collinskirui223** [comment]
I have double-checked I can see the other index files have been included I don't know how that happened. Can you reopen the PR so that I can work on the files and delete the other one?
Thanks.

**Comment by collinskirui223** [comment]
From my first commit the other index file was not included.

**Comment by Bayler** [comment]
@hectorkambow I'm not sure how that file got there. I think I'd have noticed it. How do we go about it?

**Comment by hectorkambow** [comment]
@collinskirui223 

cc @Bayler  

I think I fixed it - please double check. 

I think it was on my end. 

Confusion on my part both file names are similar (start the same) so I think I made the edits on the other - when I pushed my changes the index file was added.

**Comment by collinskirui223** [comment]
I have double-checked, and you have fixed the issue. The other index file has been removed.

**Comment by hectorkambow** [comment]
Jumping on this one

**Pull request merged**